File: tests/test_blank.js
#!/usr/bin/env node
// Blank-page detector test (JS port of test_blank.py): synthesize scan
// images with the artifacts the ADS-4300N produces on duplex backsides
// (skew wedges of dark scanner background along the edges, a fold crease,
// a light stain) and verify isBlank/inkCoverage keeps content pages and
// drops empty ones. Run: node tests/test_blank.js

const fs = require("fs");
const os = require("os");
const path = require("path");

const ocr = require(path.join(__dirname, "..", "app", "lib", "ocr.js"));

const W = 1240, H = 1754;   // A4 at 150 dpi

const failures = [];
function check(name, cond, detail = "") {
  console.log(`  [${cond ? "ok" : "FAIL"}] ${name} ${detail}`);
  if (!cond) failures.push(name);
}

const gray = (v) => `rgb(${v},${v},${v})`;

function svgPage(extra = "") {
  return Buffer.from(
    `<svg width="${W}" height="${H}" xmlns="http://www.w3.org/2000/svg">
       <rect width="${W}" height="${H}" fill="${gray(252)}"/>${extra}</svg>`);
}

const WEDGES =
  `<polygon points="0,0 ${Math.round(W * 0.45)},0 0,${Math.round(H * 0.04)}" fill="${gray(15)}"/>
   <polygon points="${W},0 ${W - Math.round(W * 0.06)},0 ${W},${Math.round(H * 0.35)}" fill="${gray(15)}"/>
   <polygon points="${W},${H} ${W - Math.round(W * 0.30)},${H} ${W},${H - Math.round(H * 0.03)}" fill="${gray(20)}"/>`;

const CREASE_AND_STAIN =
  `<line x1="0" y1="${H / 2}" x2="${W}" y2="${H / 2 - 8}"
     stroke="${gray(185)}" stroke-width="4"/>
   <ellipse cx="${W * 0.3 + 30}" cy="${H * 0.85 + 22}" rx="30" ry="22"
     fill="${gray(205)}"/>`;

const PHOTO_BLOCK =
  `<rect x="${W * 0.25}" y="${H * 0.3}" width="${W * 0.5}" height="${H * 0.25}"
     fill="${gray(70)}"/>`;

async function main() {
  const { execFileSync } = require("child_process");
  const td = fs.mkdtempSync(path.join(os.tmpdir(), "blanktest-"));
  const save = (svg, name) => {
    const p = path.join(td, name);
    fs.writeFileSync(p, execFileSync("rsvg-convert", ["--format", "png"],
                                     { input: svg, maxBuffer: 64 << 20 }));
    return p;
  };
  try {
    const clean = save(svgPage(), "clean.png");
    const wedged = save(svgPage(WEDGES + CREASE_AND_STAIN), "wedged.png");
    const photo = save(svgPage(WEDGES + PHOTO_BLOCK), "photo.png");

    const covClean = ocr.inkCoverage(clean);
    const covWedged = ocr.inkCoverage(wedged);
    const covPhoto = ocr.inkCoverage(photo);

    check("clean blank page: no ink", covClean <= 0.004,
          `(coverage ${covClean.toFixed(4)})`);
    check("skewed blank backside: wedges/crease/stain ignored",
          covWedged <= 0.004, `(coverage ${covWedged.toFixed(4)})`);
    check("photo page: interior ink detected", covPhoto > 0.004,
          `(coverage ${covPhoto.toFixed(4)})`);

    check("isBlank: empty backside with artifacts",
          (await ocr.isBlank("", wedged)) === true);
    check("isBlank: photo page with no OCR text",
          (await ocr.isBlank("", photo)) === false);
    check("isBlank: text page wins regardless of image",
          (await ocr.isBlank("Rechnung Nr. 2026-001", wedged)) === false);
    check("isBlank: missing image falls back to text",
          (await ocr.isBlank("", path.join(td, "gone.jpg"))) === true);
  } finally {
    fs.rmSync(td, { recursive: true, force: true });
  }
  console.log(`\n${failures.length} failure(s)`);
  process.exit(failures.length ? 1 : 0);
}

main().catch((e) => { console.error(e); process.exit(1); });

File: tests/test_blank.py
#!/usr/bin/python3
"""Blank-page detector test: synthesize scan images with the artifacts the
ADS-4300N produces on duplex backsides (skew wedges of dark scanner
background along the edges, a fold crease, a light stain) and verify
is_blank/ink_coverage keeps content pages and drops empty ones.
Run: /usr/bin/python3 tests/test_blank.py
"""

import os
import sys
import tempfile

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from docdoc import ocr  # noqa: E402

from PIL import Image, ImageDraw  # noqa: E402

W, H = 1240, 1754   # A4 at 150 dpi


def base_page():
    return Image.new("L", (W, H), 252)


def add_skew_wedges(im):
    """Dark scanner background peeking in at the borders of a skewed feed."""
    d = ImageDraw.Draw(im)
    d.polygon([(0, 0), (int(W * 0.45), 0), (0, int(H * 0.04))], fill=15)
    d.polygon([(W, 0), (W - int(W * 0.06), 0), (W, int(H * 0.35))], fill=15)
    d.polygon([(W, H), (W - int(W * 0.30), H), (W, H - int(H * 0.03))], fill=20)
    return im


def add_crease_and_stain(im):
    d = ImageDraw.Draw(im)
    d.line([(0, H // 2), (W, H // 2 - 8)], fill=185, width=4)   # fold shadow
    d.ellipse([(W * 0.3, H * 0.85), (W * 0.3 + 60, H * 0.85 + 45)],
              fill=205)                                         # light stain
    return im


def add_photo_block(im):
    d = ImageDraw.Draw(im)
    d.rectangle([(W * 0.25, H * 0.3), (W * 0.75, H * 0.55)], fill=70)
    return im


def main():
    failures = []

    def check(name, cond, detail=""):
        status = "ok" if cond else "FAIL"
        print(f"  [{status}] {name} {detail}")
        if not cond:
            failures.append(name)

    with tempfile.TemporaryDirectory() as td:
        def save(im, name):
            p = os.path.join(td, name)
            im.save(p, "JPEG", quality=85)
            return p

        clean = save(base_page(), "clean.jpg")
        wedged = save(add_crease_and_stain(add_skew_wedges(base_page())),
                      "wedged.jpg")
        photo = save(add_photo_block(add_skew_wedges(base_page())),
                     "photo.jpg")

        cov_clean = ocr.ink_coverage(clean)
        cov_wedged = ocr.ink_coverage(wedged)
        cov_photo = ocr.ink_coverage(photo)

        check("clean blank page: no ink", cov_clean <= 0.004,
              f"(coverage {cov_clean:.4f})")
        check("skewed blank backside: wedges/crease/stain ignored",
              cov_wedged <= 0.004, f"(coverage {cov_wedged:.4f})")
        check("photo page: interior ink detected", cov_photo > 0.004,
              f"(coverage {cov_photo:.4f})")

        check("is_blank: empty backside with artifacts",
              ocr.is_blank("", wedged) is True)
        check("is_blank: photo page with no OCR text",
              ocr.is_blank("", photo) is False)
        check("is_blank: text page wins regardless of image",
              ocr.is_blank("Rechnung Nr. 2026-001", wedged) is False)
        check("is_blank: missing image falls back to text",
              ocr.is_blank("", os.path.join(td, "gone.jpg")) is True)

    print(f"\n{len(failures)} failure(s)")
    return 1 if failures else 0


if __name__ == "__main__":
    sys.exit(main())
